ARCHITECTURE.md

```diff
@@ -0,0 +1,88 @@
+# Cargo Architecture
+
+This document gives a high level overview of Cargo internals. You may
+find it useful if you want to contribute to Cargo or if you are
+interested in the inner workings of Cargo.
+
+
+## Subcommands
+
+Cargo is organized as a set of subcommands. All subcommands live in
+`src/bin` directory. However, only `src/bin/cargo.rs` file produces an
+executable, other files inside the `bin` directory are submodules. See
+`src/bin/cargo.rs` for how these subcommands get wired up with the
```

Comment:
I'd be totally onboard with redoing how these submodules are organized as well! Something like `src/bin/subcommands/*.rs` instead of `src/bin/*.rs` if we'd like to remove this clause!